### PR TITLE
[LSP] Add in more docs for highlight groups with document_highlight()

### PR DESCRIPTION
**Pull request by ckipp01**
Currently it's not 100% clear that without setting these, using the autocomds
to utilize the `textDocument/documentHighlight` functionality, nothing will
actually be visible since the highlight groups don't have any details. This
just adds in a couple simple extra notes to make sure that's done.

Just following up on the conversation on Gitter with @clason

### CODE REVIEW

File: runtime/doc/lsp.txt
Comment (clason) [inline runtime/doc/lsp.txt:381]: Maybe add a link to the specs (https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)? I found the difference between the kinds somewhat confusing (especially `Text`).

Comment (ckipp01) [reply runtime/doc/lsp.txt:381]: Good point, I added the link.

**Review by clason — CHANGES_REQUESTED** [review]

File: runtime/doc/lsp.txt
Comment (ckipp01) [reply runtime/doc/lsp.txt:908]: Fixed.

File: runtime/doc/lsp.txt
Comment (ckipp01) [reply runtime/doc/lsp.txt:380]: Fixed.

File: runtime/doc/lsp.txt
Comment (ckipp01) [reply runtime/doc/lsp.txt:909]: I'm not sure what you mean by this. Could you explain?

Comment (clason) [reply runtime/doc/lsp.txt:909]: I meant just add a "Note:" above the "Usage" text -- which you already did, so all good 👍

**Comment by ckipp01** [comment]
I have the comments addressed, this should be ready to go!

**Review by clason — APPROVED** [review]

**Comment by ckipp01** [comment]
@clason is there a certain person I should ping to get this looked at?

**Comment by clason** [comment]
@teto @tjdevries

**Comment by ckipp01** [comment]
@mjlbach one more for you if you want 😬 

Let me know if there is anything else I can do to get this across the finish line.

File: runtime/doc/lsp.txt
Comment (mjlbach) [inline runtime/doc/lsp.txt:382]: What do you mean by `types`?

Comment (ckipp01) [reply runtime/doc/lsp.txt:382]: Well for document highlights there are three different types: 
- Text
- Read
- Write

They are explained in the given link.

Comment (mjlbach) [reply runtime/doc/lsp.txt:382]: maybe `document highlight types` for clarity, but I leave it to you if that is overly verbose

Comment (ckipp01) [reply runtime/doc/lsp.txt:382]: eh, I think it's sort of implied since that's what the section is talking about.

**Review by mjlbach — APPROVED** [review]
LGTM

**Comment by tjdevries** [comment]
Don't add doc updates in the `doc` files themselves -- you need to add them in the corresponding Lua files and then regenerate the docs `./scripts/gen_vimdocs.py` or whatever the script name is, I always forget :). At some point, we 'll add that to CI or something...

**Comment by mjlbach** [comment]
Our docs are fairly out of sync with docgen :/ I opened a PR to fix this but I think it should probably be re-run

**Comment by tjdevries** [comment]
@mjlbach which PR?

**Comment by mjlbach** [comment]
The retrospectively poorly named 'Settings validation and documentation fixes' https://github.com/neovim/neovim/pull/13698

**Comment by tjdevries** [comment]
OK, this can be rebased off of master and then we can do the documentation in the lua file and run the doc generation stuff

**Comment by ckipp01** [comment]
So I've added the changes to `buf.lua`, but the script doesn't seem to actually change anything for me. For example when I run 

```
./scripts/gen_vimdoc.py runtime/lua/vim/lsp/buf.lua
```
I see all the output on screen, but it doesn't seem to write it to disk and `gen_vimdoc.py -h` doesn't seem to show any option to persist or something. I also tried to running `gen_vimdoc.py -t lsp`, but keep getting:

```
❯ ./scripts/gen_vimdoc.py -t lsp
Traceback (most recent call last):
  File "/Users/ckipp/Documents/lua-workspace/neovim/./scripts/gen_vimdoc.py", line 1091, in <module>
    main(Doxyfile, args)
  File "/Users/ckipp/Documents/lua-workspace/neovim/./scripts/gen_vimdoc.py", line 888, in main
    p = subprocess.Popen(
  File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 947, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/local/Cellar/python@3.9/3.9.1_6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 1819, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'doxygen'
```

**Comment by mjlbach** [comment]
You need to install doxygen :P

**Comment by ckipp01** [comment]
> You need to install doxygen :P

Screenshot 2021-01-19 at 10 23 24

I have doxygen installed.

**Comment by mjlbach** [comment]
Hmm make sure it's on your path in that shell? I also believe you need messagepack installed into whatever python environment you're using.

**Comment by ckipp01** [comment]
> Hmm make sure it's on your path in that shell? I also believe you need messagepack installed into whatever python environment you're using.

Ah, that pointed me in the right direction. Thanks! Apparently the GUI and the CLI tools don't both install following the instructions on the site. This should be good to go now.

As a total aside, I'd be great to have an LSP-focused contributing section made by one of the maintainers that fully outlines how to get started, what is necessary, and odd things like, don't manually change the help docs. As someone not familiar with Neovim internals, the tools necessary to develop on it etc, I've found the process to be pretty deterring.

**Comment by teto** [comment]
Agreed, especially with the new emphasis on generating doc, we could even add a job to check if there was a manual change in a normally generated doc to prevent erronous merges as in hte past. I've created https://github.com/neovim/neovim/wiki/contribute-:help if you want to contribute your frustrations there. I will later reference it in 
 https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md

**Comment by ckipp01** [comment]
I believe this should be good to go? Is there anything else needed? I have ran doxygen like instrcutred.

**Comment by tjdevries** [comment]
it looks good to me? Can we merge? Are we waiting for anything else

**Pull request merged**
